staging/src/k8s.io/client-go/tools/cache/listwatch.go

```diff
@@ -36,6 +36,8 @@ type Lister interface {
 // Watcher is any object that knows how to start a watch on a resource.
 type Watcher interface {
 	// Watch should begin a watch at the specified version.
+	//
+	// If returning an error after the watch has started, Watch is responsible for stopping the watcher.
```

Comment:
It's just a reminder of a standard go best practice. If a function returns an error, that function should always clean up before returning or in a defer so that the caller only has to handle the error and doesn't need to handle any other cleanup caused by side effects from calling the function.

So in this case, if the Watch method implementation constructs a watch.Interface implementation and then errors later before returning, it needs to call Stop() on the watcher, because many watch.Interface implementations start go routines when the watcher is constructed, rather than waiting for ResultChan() to be called. This is a gotcha with the watch.Interface design that leaks this concern into the Watcher.Watch implementations.

If we could change the watch.Interface, we could fix this issue, but that would require a lot of work and break almost every downstream controller and tool that does watching. So instead, I opted to add comments to these interfaces to warn implementers about the implied responsibilities.

Reply:
I have extracted these comment changes to a separate PR. They don't need to block the refactor. https://github.com/kubernetes/kubernetes/pull/125472